### DISCUSSION TITLE
trino udf function issue Unknown type: array

### DISCUSSION THREAD

**Discussion by Jakin-Liu** [trinodb/trino#14978] [General]
I want to create my custom function, the function params will input array, but is always show me Unknown type: array issue.

Pls help me. 

This is my code:

```
import io.airlift.slice.Slice;
import io.trino.spi.function.Description;
import io.trino.spi.function.ScalarFunction;
import io.trino.spi.function.SqlType;
import io.trino.spi.type.StandardTypes;

import java.io.IOException;
import io.trino.spi.function.SqlNullable;

import static io.airlift.slice.Slices.utf8Slice;

public class TestFunction {

    private TestFunction() {}

    @Description("test function")
    @ScalarFunction("udf_test")
    @SqlType(StandardTypes.VARCHAR)
    public static @SqlNullable Slice udfTest (
            @SqlType(StandardTypes.VARCHAR) Slice abi,
            @SqlType(StandardTypes.VARCHAR) Slice data,
            @SqlType(StandardTypes.VARCHAR) Slice topic
    ) throws IOException {
        return utf8Slice("test trino array");
    }

    @Description("test function v2")
    @ScalarFunction("udf_test_v2")
    @SqlType(StandardTypes.VARCHAR)
    public static @SqlNullable Slice udfTestV2 (
            @SqlType(StandardTypes.VARCHAR) Slice abi,
            @SqlType(StandardTypes.VARCHAR) Slice data,
            @SqlType(StandardTypes.ARRAY) Slice topic
    ) throws IOException {
        return utf8Slice("test trino array v2");
    }
}
```

I had test code,
when I test version 1, function name is udf_test, it will show me correct result,  Please get result by pic_2:
![pic_2 pic](https://user-images.githubusercontent.com/65338514/200855278-0eaee809-ecf1-435c-ad9f-549919281e19.jpg)

When I test version 2, function name is udf_test_v2, it will show me fail result, Please get result by pic_3:
![pic_3 pic](https://user-images.githubusercontent.com/65338514/200855790-89ed78e9-1f1a-464b-8b2f-0daf3a9a6756.jpg)

for two function, the difference is that the third parameter topic's SqlType is different.

I had read docs for SqlType, I want to use StandardTypes.ARRAY for input param. But is show me fail result.

my trino version is 400, Pls help me. 
I'm very worried. Thanks

**Comment by martint** [comment]
You need to annotate the argument with `@SqlType("array(bigint)")` (or whichever concrete array type you want to support in your function), and make the type of the argument `Block`, not `Slice`.

See https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java#L1302 for an example.
